Zend/zend_hash.h

```diff
@@ -1188,6 +1188,33 @@ static zend_always_inline void *zend_hash_get_current_data_ptr_ex(HashTable *ht,
 		ZEND_HASH_FILL_FINISH(); \
 	} while (0)
 
+/* Check if an array is a list */
+static zend_always_inline zend_bool zend_array_is_list(zend_array *array)
+{
+	zend_long expected_idx = 0;
+	zend_long num_idx;
+	zend_string* str_idx;
+	/* Empty arrays are lists */
+	if (zend_hash_num_elements(array) == 0) {
+		return 1;
+	}
+
+	/* Packed arrays are lists */
+	if (HT_IS_PACKED(array) && HT_IS_WITHOUT_HOLES(array)) {
```

Comment:
Does this macro combo actually guarantee the property we are looking for? I haven't downloaded the patch, but what about this?

```php
$ht = [1, 2, 3];
array_shift($ht);
var_dump(is_list($ht));
```

I think it might satisfy being packed and without holes, but it doesn't begin at 0.

Reply:
I'm wrong, actually:

```php
$ht = [1, 2, 3];
array_shift($ht);
var_export($ht);
```

```
array (
  0 => 2,
  1 => 3,
)
```

I didn't realize it would re-order keys.